state/watcher/watcher.go

```diff
@@ -109,12 +114,16 @@ var Period time.Duration = 5 * time.Second
 
 // New returns a new Watcher observing the changelog collection,
 // which must be a capped collection maintained by mgo/txn.
-func New(changelog *mgo.Collection) *Watcher {
+func New(changelog *mgo.Collection, iteratorFunc func() mongo.Iterator) *Watcher {
 	w := &Watcher{
-		log:     changelog,
-		watches: make(map[watchKey][]watchInfo),
-		current: make(map[watchKey]int64),
-		request: make(chan interface{}),
+		log:          changelog,
+		iteratorFunc: iteratorFunc,
+		watches:      make(map[watchKey][]watchInfo),
+		current:      make(map[watchKey]int64),
+		request:      make(chan interface{}),
+	}
+	if w.iteratorFunc == nil {
```

Comment:
This is quite ick.

Couldn't you instead keep New function taking one arg, but having a test method that provider an iterator override?

Having the testing arg part of the public interface is not nice.